Support RVV f32-dwconv-unipass

yhng3010 opened the pull request:
Support f32-dwconv-unipass with `channel_tile = 8v`.

Notice that we use the variant intrinsics of tail undisturbed `tu`, 
because we observed that the original intrinsics (tail agnostic `ta`) caused register spilling problems.

Our experimental observations:
Here is the code using the original intrinsics and the code using the variant intrinsics of `_tu`:
![dw_code](https://github.com/user-attachments/assets/cba4e496-d8f0-4ba6-8d47-5f8a3d2a1328)

And here is part of the disassembly of f32-dwconv-9p8vc-minmax-rvv.c.o:
* The used toolchain is https://github.com/riscv-collab/riscv-gnu-toolchain/releases/download/2024.04.12/riscv64-glibc-ubuntu-22.04-llvm-nightly-2024.04.12-nightly.tar.gz , which is the version specified in `.github/workflows/build.yml`.
* The compiler performs a loop-unroll optimization in both cases.
* The left case have the vs8r.v instructions, which is a consequence of register spilling. The reason is that when register spilling happens, the register value is written back to memory.
* The compiler replace the vfmacc.vv with vfmadd.vv in the left case. We believe this is also an unintended consequence of register spilling.
![dw_deassembly](https://github.com/user-attachments/assets/a39601ad-8760-432b-8a4c-3bd01a474518)

yhng3010 commented:
Hi @fbarchard @bhbruce,
If you have time, can you help review this PR?

ken-unger commented:
FWIW the changes look good to me @yhng3010 and seemed to produce a good speedup when running on the K1 (vlen=256).  To address the toolchain issue that you note, would it make sense to unroll the loop with the generator?  The generated code would be a little uglier but that should avoid the compiler dependency.  Looking forward to seeing this committed.

bhbruce commented on src/f32-dwconv/unipass-rvv.c.in:
```diff
@@ -0,0 +1,76 @@
+$LMUL = int(CHANNEL_TILE[1])
+$assert KERNEL_TILE >= 2
+$assert ACTIVATION in ["LINEAR", "MINMAX"]
+$ABC = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
```
Line11: unused code.

yhng3010 replied:
Fixed. https://github.com/google/XNNPACK/pull/7402/commits/279ae034001216b1131e54419945521a24b4d676

bhbruce reviewed (approved):
Sorry for late reveiwed. Overall LGTM.
I am also curious why not to unroll the core computation using the generator.
Further more, whether the problem of ta is solved by unrolling?

BTW, I test your code on the platform with VLEN 1024.
There is an overflow bug when I ran with VLEN 1024.
https://github.com/google/XNNPACK/blob/5aa14b11b186682b542c02884dfb28fc268d2f48/src/xnnpack/config-types.h#L141-L144
In my case, channel_tile would be 256 and overflow happens.
FYR. @fbarchard

yhng3010 commented:
Hi @bhbruce @ken-unger, 

Thank you for reviewing the code.
Whether we use the specified toolchain or the latest toolchain, register spilling problem of `ta` isn't solved by unrolling.
Here is the code used for the experiment:
https://github.com/yhng3010/XNNPACK/blob/f32_dwconv_unipass_unroll_ta/src/f32-dwconv/gen/f32-dwconv-9p8vc-rvv.c

* Specified toolchain in `.github/workflows/build.yml`: https://github.com/riscv-collab/riscv-gnu-toolchain/releases/download/2024.04.12/riscv64-glibc-ubuntu-22.04-llvm-nightly-2024.04.12-nightly.tar.gz
* Latest toolchain: https://github.com/riscv-collab/riscv-gnu-toolchain/releases/download/2024.12.16/riscv64-glibc-ubuntu-22.04-llvm-nightly-2024.12.16-nightly.tar.xz

yhng3010 commented:
Hi @bhbruce,  @fbarchard

Can you help to review the PR and to take it forward?
The commit https://github.com/google/XNNPACK/pull/7402/commits/ecc9c4dfb03cbd8016e434b1b8045ccb1d15769a fix overflow in `channel_tile` and `channel_subtile`.
Since the possible maximum vlmax in RVV is 65536, the data type of tile size is now uint32_t.

ken-unger commented on src/configs/dwconv-config.c:
```diff
@@ -500,6 +500,70 @@ static void init_f32_dwconv_config(void) {
+      f32_dwconv_config[0].minmax = (xnn_dwconv_ukernel_fn) xnn_f32_dwconv_minmax_ukernel_3p8vc__rvv;
+      f32_dwconv_config[0].linear = (xnn_dwconv_ukernel_fn) xnn_f32_dwconv_ukernel_3p8vc__rvv;
+    }
+    else if (hardware_config->vlenb == 128) {
```
Should this be >=  or simply else?     There are platforms which support 512 for example.

yhng3010 replied:
This is a temporary solution, as the 4v kernel can only prevent channel_tile overflows on the platforms with vlen = 1024 (vlenb=128).
On platforms with vlen > 1024 (vlenb > 128), the 4v kernel cannot prevent channel_tile overflows.

As @bhbruce mentioned:
> BTW, I test your code on the platform with VLEN 1024. There is an overflow bug when I ran with VLEN 1024.
> 
> https://github.com/google/XNNPACK/blob/5aa14b11b186682b542c02884dfb28fc268d2f48/src/xnnpack/config-types.h#L141-L144
> 
> 
> In my case, channel_tile would be 256 and overflow happens.
> FYR. @fbarchard

The solution is to change the data type of channel_tile.
However, I'm not sure how much of an impact this change will have on the project, so I've reverted https://github.com/google/XNNPACK/commit/ecc9c4dfb03cbd8016e434b1b8045ccb1d15769a and adopted this tentative solution first.

FYR. @fbarchard @dsharlet

ken-unger replied:
Oh sorry, I was thinking vlen rather than vlenb.  Understood.

dsharlet replied:
Is the only reason you have the 4v versions to avoid this overflow?

If so, please just go ahead and change that type, and remove the extra kernels from the configs (the keeping the kernels is fine if needed, but we generally want to minimize the number of kernels used in configs).

yhng3010 replied:
Got it.
I changed the channel_tile type and removed the 4v kernels. https://github.com/google/XNNPACK/pull/7402/commits/fef4b07d9bab49771a525452af3d91c35754d4aa

yhng3010 commented:
Hi @bhbruce, @dsharlet, @fbarchard, @ken-unger 
Can you help to review the PR and to take it forward?
I have rebased the commits to the master.
- As multipass dwconv was removed, commit was changed accordingly.
- Unroll the f32-dwconv kernels
- Reverted https://github.com/google/XNNPACK/commit/ecc9c4dfb03cbd8016e434b1b8045ccb1d15769a due to lack of understanding of its impact on the project

ken-unger commented on src/f32-dwconv/unipass-rvv.c.in:
```diff
@@ -0,0 +1,73 @@
+$assert ACTIVATION in ["LINEAR", "MINMAX"]
+#include <assert.h>
+#include <riscv_vector.h>
+#include "xnnpack/dwconv.h"
```
Hi @yhng3010, unfortunately you'll need to update to latest again and regenerate as these include paths changed to "src/xnnpack/..."

yhng3010 replied:
Thank you for the notification.
The header include path has been rewritten, and the commits have been rebased onto the master.

ken-unger commented:
@dsharlet  can we get this approved PR merged.  Thank you.

dsharlet commented:
This is not compiling for us:
```

f32-dwconv/gen/f32-dwconv-4p8vc-minmax-rvv.c:64:52: error: variable 'vAcc' is uninitialized when used within its own initialization [-Werror,-Wuninitialized]
   64 |       vfloat32m8_t vAcc = __riscv_vle32_v_f32m8_tu(vAcc, w, vl);
      |                    ~~~~                            ^~~~
f32-dwconv/gen/f32-dwconv-4p8vc-minmax-rvv.c:69:37: error: variable 'va' is uninitialized when used here [-Werror,-Wuninitialized]
   69 |       va = __riscv_vle32_v_f32m8_tu(va, i0, vl);
      |                                     ^~
f32-dwconv/gen/f32-dwconv-4p8vc-minmax-rvv.c:67:7: note: variable 'va' is declared here
   67 |       vfloat32m8_t va;
      |       ^
f32-dwconv/gen/f32-dwconv-4p8vc-minmax-rvv.c:70:37: error: variable 'vb' is uninitialized when used here [-Werror,-Wuninitialized]
   70 |       vb = __riscv_vle32_v_f32m8_tu(vb, w, vl);
      |                                     ^~
f32-dwconv/gen/f32-dwconv-4p8vc-minmax-rvv.c:68:7: note: variable 'vb' is declared here
   68 |       vfloat32m8_t vb;
      |       ^
3 errors generated.
```

yhng3010 commented:
Hi @dsharlet
Sorry I didn't notice the problem, it's been fixed.
